File: 0200-number-of-islands/solution.java
class Solution {
    private void visit(char[][] grid, int i, int j) {
        grid[i][j] = '2';
        if (i > 0 && grid[i-1][j] == '1') visit(grid, i-1, j);
        if (i < grid.length - 1 && grid[i+1][j] == '1') visit(grid, i+1, j);
        if (j > 0 && grid[i][j-1] == '1') visit(grid, i, j-1);
        if (j < grid[i].length - 1 && grid[i][j+1] == '1') visit(grid, i, j+1);
    }

    public int numIslands(char[][] grid) {
        int ans = 0, m = grid.length, n = grid[0].length;
        for(int i = 0; i < m; i++) {
            for(int j = 0; j < n; j++) {
                if (grid[i][j] == '1') {
                    visit(grid, i, j);
                    ans++;
                }
            }
        }
        return ans;
    }
}

File: 0200-number-of-islands/solution.py
class Solution:
    def visitIsland(self, grid: List[List[str]], i: int, j: int, islandNumber: str):
        if i < 0 or i >= len(grid): return
        if j < 0 or j >= len(grid[0]): return
        if grid[i][j] != "1": return
        grid[i][j] = islandNumber

        self.visitIsland(grid, i+1, j, islandNumber)
        self.visitIsland(grid, i-1, j, islandNumber)
        self.visitIsland(grid, i, j+1, islandNumber)
        self.visitIsland(grid, i, j-1, islandNumber)


    def numIslands(self, grid: List[List[str]]) -> int:
        m, n, ans = len(grid), len(grid[0]), 0

        for i in range(0, m):
            for j in range(0, n):
                if grid[i][j] == "1":
                    self.visitIsland(grid, i, j, str(ans+2))
                    ans += 1
        return ans


